lib/semantics/check-omp-structure.cc

```diff
@@ -27,142 +100,168 @@ namespace Fortran::semantics {
 //                        reduction-clause |
 //                        proc-bind-clause
 void OmpStructureChecker::Enter(const parser::OmpBlockDirective::Parallel &x) {
-  std::set<std::type_index> s{typeid(parser::OmpIfClause),
-      typeid(parser::OmpClause::NumThreads), typeid(parser::OmpDefaultClause),
-      typeid(parser::OmpClause::Private),
-      typeid(parser::OmpClause::Firstprivate),
-      typeid(parser::OmpClause::Shared), typeid(parser::OmpClause::Copyin),
-      typeid(parser::OmpReductionClause), typeid(parser::OmpProcBindClause)};
-  SetAllowed(s);
-}
+  if (ompContext_.empty()) {
+    return;
+  }
+  // reserve for nesting check
+  SetOmpContextDirectiveEnum(OmpDirectiveEnum::PARALLEL);
+
+  OmpClauseEnumType allowed{OmpClauseEnum::IF, OmpClauseEnum::NUM_THREADS,
+      OmpClauseEnum::DEFAULT, OmpClauseEnum::PRIVATE,
+      OmpClauseEnum::FIRSTPRIVATE, OmpClauseEnum::SHARED, OmpClauseEnum::COPYIN,
+      OmpClauseEnum::REDUCTION, OmpClauseEnum::PROC_BIND};
+  SetOmpContextAllowed(allowed);
 
-void OmpStructureChecker::Enter(const parser::OmpBlockDirective &x) {
-  SetCurrentDirectiveSource(x.source);
+  OmpClauseEnumType allowedOnce{
+      OmpClauseEnum::IF, OmpClauseEnum::NUM_THREADS, OmpClauseEnum::PROC_BIND};
+  SetOmpContextAllowedOnce(allowedOnce);
 }
 
-void OmpStructureChecker::Leave(const parser::OmpEndBlockDirective &x) {
-  EmptyAllowed();
-  EmptyCurrentDirectiveSource();
-  EmptyCurrentClauseSource();
+// 2.7.1 do-clause -> private-clause |
+//                    lastprivate-clause |
+//                    linear-clause |
+//                    reduction-clause |
+//                    schedule-clause |
+//                    collapse-clause |
+//                    ordered-clause
+void OmpStructureChecker::Enter(const parser::OmpLoopDirective::Do &x) {
+  if (ompContext_.empty()) {
+    return;
+  }
+  // reserve for nesting check
+  SetOmpContextDirectiveEnum(OmpDirectiveEnum::DO);
+
+  OmpClauseEnumType allowed{OmpClauseEnum::PRIVATE, OmpClauseEnum::LASTPRIVATE,
+      OmpClauseEnum::LINEAR, OmpClauseEnum::REDUCTION, OmpClauseEnum::SCHEDULE,
+      OmpClauseEnum::COLLAPSE, OmpClauseEnum::ORDERED};
+  SetOmpContextAllowed(allowed);
+
+  OmpClauseEnumType allowedOnce{
+      OmpClauseEnum::SCHEDULE, OmpClauseEnum::COLLAPSE, OmpClauseEnum::ORDERED};
+  SetOmpContextAllowedOnce(allowedOnce);
 }
 
 void OmpStructureChecker::Enter(const parser::OmpClause &x) {
-  SetCurrentClauseSource(x.source);
+  if (!ompContext_.empty()) {
```

Comment:
These functions no longer use the parameter ```x```.  Would it be better to show this explicitly by removing its declaration from the function signature?  That is:
```void OmpStructureChecker::Enter(const parser::OmpClause &)```

If so, this applies to many of the functions below.

Reply:
Addressed in commit [cb89744](cb89744746186bcf120d8ca62a791e2778366b9b).